Added refs to css-text/shaping

himorin opened the pull request:
Note:

- in wpt, this might be css/css-text/shaping (so, url in @font-face need to be edited)
- need to add [Noto Naskh Arabic webfont](https://www.google.com/get/noto/#naskh-arab) to wpt
- no reference applicable to exp-sub-grapheme-highlight-000.html??

himorin committed:
Possible fix to shaping-014, wrong font src (ref), wrong match href (test)

himorin committed:
added ref for shaping-000

himorin committed:
fixed shaping-014 to pass wpt lint

himorin committed:
swapped all tabs to white spaces, removed font from flags

himorin committed:
removed blank meta/flags line, added rel=match line

himorin committed:
fixed meta:flags, tab to wsp

himorin committed:
references/shaping-000: back div from ref to test (no ref class definition)

himorin committed:
added shaping-000-ref as finished one

himorin committed:
commit shaping-014-ref

himorin committed:
added -ref files copied from original test with removing meta/link lines

himorin committed:
Added U+200D for linking U+200C for not linking to references

himorin commented:
one possible point which could be pointed by wpt: wpt reftest uses 600x600 window size, some test with two boxex seems overflowing for that size as for now.

r12a commented:
@himorin i see a couple of issues straight away:

1. the ref file for shaping-001-ref.html doesn't look as you would expect, so we can't use that approach for the reference file in that case.  That suggests to me that you didn't actually look at the tests and refs to check that they work as expected - that's always an important thing to do.

2. it's quite difficult to review/debug the tests when all the joiner characters are invisible - please use &amp;#x200D; rather than the code point in the source text.

We need to find an alternative approach for the ref files, if possible. If we can't then we'll have to submit the file as a manual test.  As long as we control the font, we may be able to do something with the Arabic presentation forms in Unicode. I'll take a look.

r12a commented on css-text/shaping/reference/exp-sub-grapheme-highlight-000-ref.html:
```diff
@@ -0,0 +1,111 @@
+<!DOCTYPE html>
```
exp-sub-grapheme-highlight-000.html is an exploratory test, so not appropriate for WPT. We also don't need a ref file for it.  So we should remove this file from the set.

himorin committed:
replaced all zwj/zwnj to reference format

himorin committed:
removed exp-sub-grapheme-highlight-000-ref.html

himorin committed:
added dir="rtl" to all

himorin committed:
switched U+200D to reference (&zwj;)

himorin committed:
added .ref style, changed from test to ref for reference, added second box for reference (except cchar, lig-012)

himorin committed:
switched font-size to 120px, considering 2nd panel for -008 which have 120% for testing and isolated shape.

vim commands
:args s*html reference/s*html
:argdo %s/font-size: 172px; /font-size: 120px; /
:argdo wq

himorin commented:
@r12a 
I've modified:
- added dir="rtl" to all divs for test/ref
- changed font size to 120px, but not 150px. Its to deal with some tests which have highter box, that some test will show isolated shape (which has highter shape than tripled ones used for tests which we've examined font size) and also three -008 tests uses 120% for middle.
- added style .ref to all files (the same definition as .test)
- added visual references to tests with only one box - with class of test/ref for test, ref/ref for reference, some tests which has two boxes for test target (all cchar, lig-012) have not modified

Remainings
- how to do for all cchar and lig-012, to add visual reference

r12a committed:
Merge branch 'gh-pages' into himorin-css-text-shaping-refs

r12a committed:
Merge remote-tracking branch 'origin/himorin-css-text-shaping-refs' into himorin-css-text-shaping-refs

r12a commented:
@himorin I ran through the tests on Firefox, Chrome & Safari (on Mac) and spotted these problems.  cc @xfq @frivoal @fantasai for FYI or in case you have ideas.

- shaping-015 and a couple more: Instructions say "Test passes if both boxes look the same." - i mentioned this yesterday on the phone, and i think i suggested that this would be better wording, however i think i was wrong.  I think the two boxes are a good enough clue that this is a ref test, and it's more useful to say what the criterion is for testing. The good news is that those particular instructions are only set on a handful of test files, so it should be easy to change them

- shaping 2 and 18:  the ref text has a gap between the first and second characters in FF - we need to investigate why. As it is, that will cause the test to fail incorrectly.  If we can't figure out how to fix the ref test, we'd need to revert these tests to manual.

- the ref for shaping 12 is incorrect in FF - it joins when it shouldn't.  Again, this is a deal breaker.

- the ref for shaping 20 on Chrome doesn't join - it may be a question of using a Noto Sans N'Ko webfont.

- shaping 23-25 refs fail because Safari doesn't support vertical mongolian properly - we probably need to revert to manual for those tests.

- shaping-cchar 0 - 8: the ref box text doesn't join for Chrome or Safari. Need to investigate whether there's an alternative approach here.  

- shaping-cchar 9-11: the ref box text joins in FF when it shouldn't

- shaping-lig all: there are 2 possible outcomes for these tests, so a ref test is inappropriate.

Could you investigate those issues and recommend to me the best way forward.

r12a commented:
For the benefit of those i just cc'ed, the current version of these tests is available (without @himorin's changes) from https://w3c.github.io/i18n-tests/results/css-text-shaping

himorin commented:
- For shaping 14-18, let me change "Test passes if both boxes look the same." (actually, it's the original) to "JOIN" or "DON'T JOIN".
- For ref of shaping [2](https://i18n-tests-w3c.himor.in/css-text/shaping/reference/shaping-002-ref.html), [18](https://i18n-tests-w3c.himor.in/css-text/shaping/reference/shaping-018-ref.html), and [20](https://i18n-tests-w3c.himor.in/css-text/shaping/reference/shaping-020-ref.html), I could not reproduce on FF or Chrome.
- For ref of shaping 12 on FF, I suppose a possible fix is to add zwnj for 2nd box. Looking like [this file](https://i18n-tests-w3c.himor.in/css-text/shaping/reference/shaping-009-test.html), it seems FF apply separation from span/css after applying shape (would not reproduce on FF nightly, thanks to Masayuki from Mozilla on pointing difference)??
- For ref of shaping 23-25, sorry that I cannot test using Safari, let me make this back to manual.
- For cchar 0-11, as I put in remaining points of previous comment, original test has two different boxes (with FATHA and SHADDA), so I'd hear what is a better way to add references? (adding two more boxes could be possible, but could be difficult to tell which pairs to be compared. dividing into two test cases could be another possibility.)
- For cchar 9-11, again, working for me...
- For lig all, could you let me learn what "2 possible outcomes" means? are there any different shaping when characters "DON'T join"??

r12a commented:
> I could not reproduce 

My testing was done on a Mac. Are you able to get access to a Mac when you travel to the office?  It's a bit difficult for you to work on tests if you can't run Safari...

To speed things up, i made the changes.

shaping-015 and a couple more:  fixed

shaping 2 and 18: I left this as is, since FF produces a fail anyway, and if it didn't perhaps we would see the same gap anyway.  I can't find anything wrong with the test.

the ref for shaping 12: The problem was that there was no ref for this test (there are 2 test boxes), and the second had been incorrectly turned into a ref box in the ref file. Fixed by turning it into 2 tests.

ref for shaping 20 on Chrome: Seems to be working now.

shaping 23-25 refs fail because Safari doesn't support vertical: remove the ref tests and divs

shaping-cchar 0 - 8: &
shaping-cchar 9-11: Again, 2nd tests were being regarded as refs. I merged the 2 tests into one and added refs for all. Problem solved.

shaping-lig all: These are all exploratory tests; it's not clear what should happen from the spec, and it's even possible that more than one thing could happen.  I removed the ref divs and tests.

I pushed the new set of files directly to github, so i'll close this PR.  Please synch your local copy with the latest. (I had planned to upload your changes so far first, so you could see the things i changed, but i managed to spoil my local copy of your files, then couldn't find your fork, so i went ahead with all changes to date.)

r12a commented:
The commit history for the the upload i just did is at https://github.com/w3c/i18n-tests/commit/f3d9191be20543c63bb1022f5c5fc1c359002861

r12a commented:
@himorin We need to also add the N'Ko and Mongolian webfonts to WPT fonts/noto directory. Could you raise a pull request for that?
